packages/overseer-rs/src/ethereum/mod.rs

```diff
@@ -0,0 +1,207 @@
+use std::{
+    collections::HashMap,
+    time::{Duration, SystemTime},
+};
+
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+use reqwest::Client;
+use serde::de::DeserializeOwned;
+use serde::Deserialize;
+use tracing::debug;
+
+use crate::types::{Observation, PendingTransaction};
+
+/// Minimal snapshot of a block required for blacklist evaluations.
+#[derive(Clone, Debug)]
+pub struct BlockSnapshot {
+    pub number: u64,
+    pub timestamp: SystemTime,
+    pub transaction_count: u64,
+}
+
+/// Client abstraction for querying Ethereum execution data.
+#[async_trait]
+pub trait EthereumClient: Send + Sync {
+    /// Returns the most recent block snapshot from the execution layer.
+    async fn latest_block(&self) -> Result<BlockSnapshot>;
+
+    /// Returns the set of pending transactions currently in the mempool.
+    async fn pending_transactions(&self) -> Result<Vec<PendingTransaction>>;
+}
+
+/// JSON-RPC backed implementation of the [`EthereumClient`] trait.
+#[derive(Clone, Debug)]
+pub struct RpcEthereumClient {
+    http: Client,
+    rpc_url: String,
+}
+
+impl RpcEthereumClient {
+    /// Constructs a client that targets the supplied RPC endpoint.
+    pub fn new(rpc_url: impl Into<String>) -> Self {
+        Self {
+            http: Client::new(),
+            rpc_url: rpc_url.into(),
+        }
+    }
+
+    /// Issues a typed JSON-RPC request and deserialises the result.
+    async fn rpc_call<T>(&self, method: &str, params: serde_json::Value) -> Result<T>
+    where
+        T: DeserializeOwned,
+    {
+        let payload = serde_json::json!({
+            "jsonrpc": "2.0",
+            "id": 1,
+            "method": method,
+            "params": params,
+        });
+
+        let response = self
+            .http
+            .post(&self.rpc_url)
+            .json(&payload)
+            .send()
+            .await
+            .with_context(|| format!("RPC request to {} failed", self.rpc_url))?;
+
+        let status = response
+            .error_for_status()
+            .with_context(|| format!("RPC request to {} returned error status", self.rpc_url))?;
+
+        let body: RpcResponse<T> = status
+            .json()
+            .await
+            .with_context(|| "failed to decode RPC response body".to_string())?;
+
+        if let Some(error) = body.error {
+            return Err(anyhow!(
+                "RPC error {} (method {}): {}",
+                error.code,
+                method,
+                error.message
+            ));
+        }
+
+        body.result
+            .ok_or_else(|| anyhow!("missing result field in RPC response"))
+    }
+}
+
+#[async_trait]
+impl EthereumClient for RpcEthereumClient {
+    /// Fetches the latest block metadata using `eth_getBlockByNumber`.
+    async fn latest_block(&self) -> Result<BlockSnapshot> {
+        let block: RpcBlock = self
+            .rpc_call("eth_getBlockByNumber", serde_json::json!(["latest", false]))
```

Comment:
I wonder why don't you use alloy for accessing Ethereum RPC. E.g. block can be read with `get_block_by_number` from alloy provider: https://github.com/NethermindEth/Catalyst/blob/master/common/src/shared/execution_layer.rs#L98